trustymail/trustymail.py

```diff
@@ -328,12 +318,104 @@ def dmarc_scan(resolver, domain):
+                    values = tag_dict[tag].split(':')
+                    for value in values:
+                        if value not in ['afrf']:
+                            msg = 'Unknown DMARC rf tag value {0}'.format(value)
+                            handle_syntax_error('[DMARC]', domain, '{0}'.format(msg))
+                            domain.valid_dmarc = False
+                elif tag == 'ri':
+                    try:
+                        int(tag_dict[tag])
+                    except ValueError:
+                        msg = 'Invalid DMARC ri tag value: {0} - must be an integer'.format(tag_dict[tag])
+                        handle_syntax_error('[DMARC]', domain, '{0}'.format(msg))
+                        domain.valid_dmarc = False
+                elif tag == 'pct':
+                    try:
+                        pct = int(tag_dict[tag])
+                        if pct < 0 or pct > 100:
+                            msg = 'Error: invalid DMARC pct tag value: {0} - must be an integer between ' \
+                                  '0 and 100'.format(tag_dict[tag])
+                            handle_syntax_error('[DMARC]', domain, '{0}'.format(msg))
+                            domain.valid_dmarc = False
+                        domain.dmarc_pct = pct
+                    except ValueError:
+                        msg = 'invalid DMARC pct tag value: {0} - must be an integer'.format(tag_dict[tag])
+                        handle_syntax_error('[DMARC]', domain, '{0}'.format(msg))
+                        domain.valid_dmarc = False
+                elif tag == 'rua' or tag == 'ruf':
+                    uris = tag_dict[tag].split(',')
+                    for uri in uris:
+                        # mailto: is currently the only type of DMARC URI
+                        mailto_matches = MAILTO_REGEX.findall(uri)
+                        if len(mailto_matches) != 1:
+                            msg = 'Error: {0} is an invalid DMARC URI'.format(uri)
+                            handle_syntax_error('[DMARC]', domain, '{0}'.format(msg))
+                            domain.valid_dmarc = False
+                        else:
+                            email_address = mailto_matches[0]
+                            email_domain = email_address.split('@')[-1]
+                            if email_domain.lower() != domain.domain_name.lower():
+                                target = '{0}._report._dmarc.{1}'.format(domain.domain_name, email_domain)
+                                error_message = '{0} does not indicate that it accepts DMARC reports about {1} - ' \
+                                                'https://tools.ietf.org' \
+                                                '/html/rfc7489#section-7.1'.format(email_domain,
+                                                                                   domain.domain_name)
+                                handle_error('[DMARC]', domain, '{0}'.format(error_message))
```

Comment:
Do we want this line here?  I think we only want to publish the error if we detect an error condition below, right?

Reply:
I think this code is correct according to the standard, but a ton of domains will fail this check.  Many domains include an email address of their own as a secondary collection location for rua and/or ruf messages, which often makes this check fail.  For instance, `dhs.gov` and `nasa.gov` both fail because of this check.

I'm not sure what the right thing is to do here...treat this as an error or treat it as a warning?  Any insight, @h-m-f-t?

Reply:
I struggled with the error vs warning decision in my own project. My checker is for checking if a DMARC record, and the issue needs to be fixed by the managers of the receiving end domain, who may not be the same people who deploy the DMARC record on the other domain. However...

For DHS, the spirit of the BOD and trustymail is to permote and validate best practices. Without these records in place, they will not receive reports, and could be under the false impression that everything is fine when they move to regect later next year.

If I remember correctly, I also made this an error because my understanding of issue #33 was that @h-m-f-t calls for this condition to invalidate the record.

Related thoughts:

- Is it possible to amend the BOD to include information and instructions about this additional requirement, and require that pct is not set to less than 100 (100 is the implicit default if pct is not specified)
- I didn't notice a field for warnings in the CSV/JSON. I think that should be added. Otherwise, they could be easily overlooked. Do they currently only show up in the console and only in debug?
- Having the module name and full file path added to each warning or error makes the CSV field very wide and difficult to read, even on a ultrawide monitor. Can you remove the excess data, and only use it in --debug console output?

Reply:
There were two questions (confusingly hidden) in my comments above.  You spoke to the second one.  The first one, though, I think is still relevant.  I think line 402 should be removed, since we don't want to report an error (or warning) until we actually encounter one.  I hope that makes sense.

Reply:
Regarding your three bullet points:
* Probably, but that's a question for @h-m-f-t.
* I'm not against a warnings field.  Right now syntax errors go to "Errors" and everything else (including warnings) go to "Debug Info", all via `handle_error()` in `trustymail.py`.
* We could remove the irrelevant parts of the path in `handle_error()`.  Would that help?  I like having the extra information there because we can go back and dig around in the CSV output when folks receive their report and want to know why we gave them the results we did, but I agree that what is there now is a bit cumbersome.  I also think that JSON is a better output format for this data then CSV, but we want to have a format that won't frighten away non-technical users.

Reply:
You are right about line 402. I totally missed that copy/paste error. Nice catch!  Removed.

Reply:
I like have the extra info to track down why errors failed too. Maybe only output the extra stuff in JSON so you keep the user-friendly CSV format concise, and put in the docs somewhere to output JSON if they want more details?

Reply:
> I think this code is correct according to the standard, but a ton of domains will fail this check

@jsf9k, yup, this is exactly the behavior we want. BOD 18-01 requires agencies have "[valid](https://cyber.dhs.gov/#ii-required-actions)" records, and I would argue that if you configure something that isn't actually going to work, that's invalid.

> Is it possible to amend the BOD to include information and instructions about this additional requirement, and require that pct is not set to less than 100 (100 is the implicit default if pct is not specified)

@seanthegeek, no, it's not really possible to update the BOD itself, but it's definitely possible to include this information in the [implementation guide](https://cyber.dhs.gov/guide/) and then require it as part of our scanning regime-- and that's what we'll do. 
>Having the module name and full file path added to each warning or error makes the CSV field very wide and difficult to read, even on a ultrawide monitor. Can you remove the excess data, and only use it in --debug console output?

@seanthegeek, I agree that having the module/path is noisy, but I would like to retain the rest of the info in the CSV.

Reply:
@jsf9k @h-m-f-t I think only using the filename,  rather than the full path to the file in `handle_error()` is a good solution.